### PR TITLE
refactor: switch to NPM workspaces

### LINKED ISSUE #685: ci: dependencies CI job is sometimes failing
The `lerna bootstrap` command part of the `dependencies` CI job is currently sometimes failing with the error message:

```
npm ERR! code 1
npm ERR! path /home/circleci/axe-core-npm/packages/webdriverio/node_modules/chromedriver
npm ERR! command failed
npm ERR! command sh -c node install.js
npm ERR! Current existing ChromeDriver binary is unavailable, proceeding with download and extraction.
npm ERR! Downloading from file:  https://chromedriver.storage.googleapis.com/109.0.5414.74/chromedriver_linux64.zip
npm ERR! Saving to file: /tmp/109.0.5414.74/chromedriver/chromedriver_linux64.zip
npm ERR! Received 1040K...
npm ERR! Received 2080K...
npm ERR! Received 3120K...
npm ERR! Received 4160K...
npm ERR! Received 5200K...
npm ERR! Received 6240K...
npm ERR! Received 7129K total.
npm ERR! Extracting zip contents to /tmp/109.0.5414.74/chromedriver.
npm ERR! ChromeDriver installation failed Error: Error extracting archive: Error: unexpected end of file
npm ERR!     at extractDownload (/home/circleci/axe-core-npm/packages/webdriverio/node_modules/chromedriver/install.js:309:11)
npm ERR!     at async install (/home/circleci/axe-core-npm/packages/webdriverio/node_modules/chromedriver/install.js:71:7)
```

This appears to be some unreliable CI fluke since the job failed and succeeded once [for the exact same commit](https://app.circleci.com/pipelines/github/dequelabs/axe-core-npm?branch=pull%2F684):

image

I am not sure what exactly is causing this CI failure ... it doesn't happen locally for me. Changing the command to `npm run bootstrap -- --ci || (md5sum /tmp/109.0.5414.74/chromedriver/chromedriver_linux64.zip; ls -l /tmp/109.0.5414.74/chromedriver; false)` [revealed that the zip extraction indeed is successful](https://app.circleci.com/pipelines/github/dequelabs/axe-core-npm/1525/workflows/8d80f53e-17b9-4a5d-91da-b7e3c609458c/jobs/12559) ... which makes the occurrence of the "Error extracting archive" all the more puzzling.

Inserting a `cd packages/webdriverio && npm ci` before the `lerna bootstrap` call fixed the problem, showing that this indeed appears to be a lerna bug. Since `lerna bootstrap` is [deprecated anyway](https://github.com/lerna/lerna/discussions/3410) it probably makes sense to switch to NPM workspaces.

I have two PRs that are affected by this:

* #681 [CircleCI](https://app.circleci.com/pipelines/github/dequelabs/axe-core-npm/1518/workflows/78e54309-10f2-48ae-beac-da2a1318c6f2/jobs/12516)
* #684 [CircleCI](https://app.circleci.com/pipelines/github/dequelabs/axe-core-npm/1522/workflows/8e18eaf2-f4b1-459f-9421-333e7ed0e075/jobs/12547)

### PR DESCRIPTION
**Pull request by not-my-profile**
The `lerna bootstrap` command is deprecated[1]
and resulted in a very weird CI failure (#685).

This commit switches the repository to instead use
NPM workspaces.[2]

[1]: https://github.com/lerna/lerna/discussions/3410
[2]: https://docs.npmjs.com/cli/v7/using-npm/workspaces

Fixes #685.

### CODE REVIEW

File: packages/reporter-earl/tests/axeReporterEarl.test.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { getDummyData } from './utils';
 import axeReporterEarl, { createEarlReport } from '../src/axeReporterEarl';
 import * as context from '../src/context.json';
 import { RawResult, EarlType } from '../src/types';
+import { describe, beforeEach, test, expect } from '@jest/globals';
```
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply packages/reporter-earl/tests/axeReporterEarl.test.ts:7]: It looks like the build is failing due to Jest related dependency issues.

[the remark this answers is not in the document]
Comment (stephenmathieson) [reply packages/reporter-earl/tests/axeReporterEarl.test.ts:7]: That's odd. I was not aware of that. Thanks 👍

**Comment by not-my-profile** [comment]
Ok I got it working ... the failing playwright test is just a fluke, see #689. I think we should disable that test until we figure that out.

I split off the first two commits into a separate PR: #690, so they can be independently reviewed.

**Comment by not-my-profile** [comment]
This is ready for review ... I don't know what's causing the playwright CI fluke, see #689 ... but I also cannot easily debug that since the CIs are currently almost always failing at the prior dependencies job (#685), which is fixed by this PR.

File: .circleci/config.yml [hunk 2]
```diff
@@ -187,6 +128,7 @@ jobs:
       - restore_dependency_cache
       - browser-tools/install-chrome
       - browser-tools/install-chromedriver
+      - run: npm install -g npm@7
```
Comment (stephenmathieson) [inline .circleci/config.yml:131]: Why pin this at v7? I imagine this will be easy to overlook when we upgrade Node.js.

File: packages/cli/package.json [hunk 3]
```diff
@@ -32,8 +32,7 @@
     "prebuild": "rimraf dist",
     "build": "tsc",
     "test": "mocha --timeout 60000 -r ts-node/register 'src/**/**.test.ts'",
-    "coverage": "nyc npm run test",
-    "prepare": "npm run build"
```
Comment (stephenmathieson) [inline packages/cli/package.json:36]: Why is the `prepare` script being removed? IIRC we rely on it to transpile before publishing to npm.

File: packages/cli/src/lib/utils.test.ts [hunk 4]
```diff
@@ -73,25 +72,6 @@ describe('utils', () => {
...
-          '../node_modules/axe-core': {
-            'axe.js': mock.load(require.resolve('axe-core'))
-          }
-        });
-      });
-
-      afterEach(() => {
-        mock.restore();
-      });
-      it('fall back to use `locally` installed axe-core', () => {
```
Comment (stephenmathieson) [inline packages/cli/src/lib/utils.test.ts:89]: Can we implement a different test rather than removing this?

File: packages/cli/src/lib/utils.ts [hunk 5]
```diff
@@ -82,23 +82,7 @@ export const getAxeSource = (axePath?: string): string | void => {
   }
 
   if (!fs.existsSync(axePath)) {
-    // Look for axe in CWD ./node_modules
-    axePath = path.join(process.cwd(), 'node_modules', 'axe-core', 'axe.js');
```
Comment (stephenmathieson) [inline packages/cli/src/lib/utils.ts:86]: Should this be considered a breaking change? @michael-siek

**Review by stephenmathieson — COMMENTED** [review]
Few small things left inline. This is looking good thought 👍 

Will let someone with more recent familiarity with this project give the final review.

File: .circleci/config.yml (hunk 2 above)
Comment (not-my-profile) [reply .circleci/config.yml:131]: A few lines above that we we're using a node14 image due to WDIO. Node 14 comes with npm 6.14, which doesn't yet support NPM workspaces.

File: packages/cli/src/lib/utils.test.ts (hunk 4 above)
Comment (not-my-profile) [reply packages/cli/src/lib/utils.test.ts:89]: The fact that it defaults to some locally installed axe-core is already tested indirectly through several other test cases. The node_module resolution code has been simplified to `axePath = require.resolve('axe-core/axe.js');`, which I don't really think requires dedicated testing.

**Comment by michael-siek** [comment]
Hey @not-my-profile just reaching out to see if you've seen the request changes and merge conflicts on this PR.

**Comment by not-my-profile** [comment]
I rebased the PR but I currently don't have time to look into the new CI failures. I split off the first four preparatory commits into #757 ... I think it makes sense to land these changes first.

**Comment by Zidious** [comment]
> I rebased the PR but I currently don't have time to look into the new CI failures. I split off the first four preparatory commits into #757 ... I think it makes sense to land these changes first.

Hey @not-my-profile, 

When you get a moment, there are merge conflicts. Feel free to close and re-open if you don't have time at the moment. 

Thanks again!

**Comment by not-my-profile** [comment]
If I close it, I'll probably forget that it exists. I just converted it to a draft to signal that it's not ready for review yet (due to the CI failures I mentioned earlier).

File: packages/cli/package.json [hunk 6]
```diff
@@ -45,7 +44,7 @@
   ],
   "dependencies": {
     "@axe-core/webdriverjs": "^4.7.3",
-    "axe-core": "^4.7.0",
+    "axe-core": "~4.7.0",
```
Comment (straker) [inline packages/cli/package.json:47]: Axe-core 4.8.1 was released and I don't want to take it with this conversion, so pinned all axe-core versions to 4.7

File: packages/cli/src/bin/cli.test.ts [hunk 7]
```diff
@@ -391,7 +391,7 @@ describe('cli', () => {
     });
   });
 
-  describe('--chromedriver-path', () => {
+  describe.skip('--chromedriver-path', () => {
```
Comment (straker) [inline packages/cli/src/bin/cli.test.ts:394]: This test fails due to the way workspaces manages node_modules, will open a tech debt ticket to fix it.

File: packages/puppeteer/package.json [hunk 8]
```diff
@@ -44,7 +44,7 @@
     "express": "^4.18.2",
     "mocha": "^10.0.0",
     "nyc": "^15.1.0",
-    "puppeteer": "^19.7.3",
+    "puppeteer": "19.7.3",
```
Comment (straker) [inline packages/puppeteer/package.json:47]: Starting in Puppeteer 19.11 (ish?) they started adding a log that stated that `headlness: new` would be needed (which massively clogs up the test run as it logs a huge paragraph every test). However, if you add `headless: new` to the options it breaks the unloaded iframe test. It seems that whatever is in `headless: new` fixes the problem with unloaded iframes? Will open a tech debt ticket to investigate.

File: packages/webdriverio/package.json [hunk 9]
```diff
@@ -70,7 +70,7 @@
     "ts-node": "^10.9.1",
     "tsup": "^6.7.0",
     "typescript": "^4.8.4",
-    "webdriverio": "^8.8.2"
+    "webdriverio": "8.8.2"
```
Comment (straker) [inline packages/webdriverio/package.json:73]: Latest webdriverio breaks our typescript build due to this bug with types https://github.com/webdriverio/webdriverio/issues/11055. Will open a tech debt ticket to fix it.

File: packages/cli/package.json (hunk 3 above)
Comment (straker) [inline packages/cli/package.json:36]: Talked with @michael-siek and this isn't needed. When left in it tries to build the cli before it links packages together, which would cause the build to fail as @axe-core/webdriverjs wasn't available yet.

File: packages/puppeteer/tsconfig.json [hunk 10]
```diff
@@ -1,24 +1,4 @@
...
-    "removeComments": true /* Do not emit comments to output. */,
-    "pretty": true,
-    "strict": true,
-    "noImplicitAny": true,
-    "noImplicitThis": true,
-    "noImplicitReturns": true,
-    "moduleResolution": "node",
-    "esModuleInterop": true
-  },
+  "extends": "../../tsconfig.json",
```
Comment (dbjorge) [inline packages/puppeteer/tsconfig.json:2]: This is hiding a change from `module: commonjs` to `module: esnext` that probably shouldn't be part of this PR

File: packages/react/examples/next.js/package-lock.json
Comment (dbjorge) [inline packages/react/examples/next.js/package-lock.json:1]: Should these examples be added as workspaces so CI will verify they build?

**Review by dbjorge — CHANGES_REQUESTED** [review]

File: packages/react/examples/next.js/package-lock.json
Comment (straker) [reply packages/react/examples/next.js/package-lock.json:1]: We never had tests to run the examples so they were never verified to begin with. If we want to have them run tests we should do so in a different pr.

File: packages/puppeteer/package.json [hunk 11]
```diff
@@ -53,10 +53,7 @@
     "typescript": "^4.8.4"
   },
   "dependencies": {
-    "axe-core": "^4.7.0"
-  },
-  "peerDependencies": {
```
[the remark this answers is not in the document]
Comment (straker) [reply packages/puppeteer/package.json:58]: It was not. Not sure when that happened. Good catch.

**Review by dbjorge — APPROVED** [review]

**Comment by dbjorge** [comment]
force-push was to correct author email address in merge from develop, to hopefully fix CLA bot

**Comment by dbjorge** [comment]
Reviewed for security

**Review by dbjorge — DISMISSED** [review]
Ah, wait, approved too early - @straker , there's a missing lockfile update to pick up the changes from 2dedf1b

**Review by michael-siek — APPROVED** [review]

File: packages/puppeteer/package.json (hunk 8 above)
Comment (straker) [reply packages/puppeteer/package.json:47]: https://github.com/dequelabs/axe-core-npm/issues/828

File: packages/webdriverio/package.json (hunk 9 above)
Comment (straker) [reply packages/webdriverio/package.json:73]: https://github.com/dequelabs/axe-core-npm/issues/829

**Comment by michael-siek** [comment]
- [x] reviewed for security
